File: inst/extdata/convert-bcsd-v3.py
"""
Convert bcsd.zarr (V2) to bcsd_v3.zarr (V3).

Unzips the V2 bcsd.zarr.zip, reads each array and group,
and writes an equivalent Zarr V3 store using zarr-python >= 3.x.

Usage:
  pip install "zarr>=3" numpy numcodecs
  python convert-bcsd-v3.py

Outputs bcsd_v3.zarr.zip in the same directory as this script.
"""

import os
import shutil
import tempfile
import zipfile

import numpy as np
import zarr

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
V2_ZIP = os.path.join(SCRIPT_DIR, "bcsd.zarr.zip")
V3_ZIP = os.path.join(SCRIPT_DIR, "bcsd_v3.zarr.zip")

# Clean up any previous output
if os.path.exists(V3_ZIP):
    os.remove(V3_ZIP)

# Work entirely in a temp directory to avoid Windows permission issues
tmpdir = tempfile.mkdtemp()
v2_extract = os.path.join(tmpdir, "v2")
v3_dir = os.path.join(tmpdir, "bcsd_v3.zarr")

# Unzip V2 store
with zipfile.ZipFile(V2_ZIP, "r") as zf:
    zf.extractall(v2_extract)

# The zip contains files under bcsd.zarr/
v2_path = os.path.join(v2_extract, "bcsd.zarr")
print(f"V2 store extracted to: {v2_path}")

v2_root = zarr.open_group(v2_path, mode="r", zarr_format=2)

# Create V3 output store in temp directory
v3_root = zarr.open_group(v3_dir, mode="w", zarr_format=3)

# Copy root-level attributes (use update_attributes for single write)
root_attrs = dict(v2_root.attrs)
v3_root = v3_root.update_attributes(root_attrs)
print(f"Root attributes: {len(root_attrs)} keys copied")

# Copy each array
for name in v2_root.array_keys():
    v2_arr = v2_root[name]
    data = v2_arr[:]

    print(f"  Array '{name}': shape={v2_arr.shape}, dtype={v2_arr.dtype}, "
          f"chunks={v2_arr.chunks}")

    # Create V3 array then write data separately
    v3_arr = v3_root.create_array(
        name,
        shape=v2_arr.shape,
        chunks=v2_arr.chunks,
        dtype=v2_arr.dtype,
        fill_value=v2_arr.fill_value,
    )
    v3_arr[:] = data

    # Copy array attributes
    arr_attrs = dict(v2_arr.attrs)
    if arr_attrs:
        v3_arr.update_attributes(arr_attrs)
        print(f"    Attributes: {list(arr_attrs.keys())}")

print(f"\nV3 store written to: {v3_dir}")

# Verify by reading back
v3_check = zarr.open_group(v3_dir, mode="r")
for name in v3_check.array_keys():
    arr = v3_check[name]
    print(f"  Verify '{name}': shape={arr.shape}, dtype={arr.dtype}")

# Zip it up directly to the final location
print(f"\nCreating {V3_ZIP}...")
with zipfile.ZipFile(V3_ZIP, "w", zipfile.ZIP_DEFLATED) as zf:
    for root, dirs, files in os.walk(v3_dir):
        for f in files:
            full_path = os.path.join(root, f)
            arc_name = os.path.relpath(full_path, os.path.dirname(v3_dir))
            zf.write(full_path, arc_name)

print("Done!")

# Clean up temp directory
shutil.rmtree(tmpdir)
print("Cleaned up temp directories")

File: inst/extdata/convert-bcsd-v3.R
# Convert bcsd_v2.zarr.zip (V2) to bcsd_v3.zarr.zip (V3) using pizzarr.
#
# R equivalent of convert-bcsd-v3.py.
# Unzips the V2 store, reads each array and group,
# and writes an equivalent Zarr V3 store, then zips it.
#
# Usage:
#   Rscript convert-bcsd-v3.R

library(pizzarr)

get_script_dir <- function() {
  # Works with Rscript --file, source(), and commandArgs
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- grep("^--file=", args, value = TRUE)
  if (length(file_arg) > 0) {
    return(dirname(normalizePath(sub("^--file=", "", file_arg))))
  }
  # Fallback: assume working directory contains inst/extdata
  return(normalizePath("inst/extdata"))
}
script_dir <- get_script_dir()

v2_zip <- file.path(script_dir, "bcsd_v2.zarr.zip")
v3_zip <- file.path(script_dir, "bcsd_v3.zarr.zip")

stopifnot(file.exists(v2_zip))

# Clean up previous output
if (file.exists(v3_zip)) file.remove(v3_zip)

# Work in temp directories
tmpdir <- tempfile("bcsd_convert")
dir.create(tmpdir)
on.exit(unlink(tmpdir, recursive = TRUE), add = TRUE)

v2_extract <- file.path(tmpdir, "v2")
v3_dir <- file.path(tmpdir, "bcsd_v3.zarr")

# Unzip V2 store
utils::unzip(v2_zip, exdir = v2_extract)
v2_path <- file.path(v2_extract, "bcsd_v2.zarr")
cat("V2 store extracted to:", v2_path, "\n")

# Open V2 root group
v2_root <- zarr_open_group(v2_path, mode = "r")

# Create V3 output store
dir.create(v3_dir)
v3_store <- DirectoryStore$new(v3_dir)
v3_root <- zarr_create_group(store = v3_store, zarr_format = 3L)

# Copy root-level attributes
root_attrs <- v2_root$get_attrs()$to_list()
if (length(root_attrs) > 0) {
  v3_attrs <- v3_root$get_attrs()
  for (nm in names(root_attrs)) {
    v3_attrs$set_item(nm, root_attrs[[nm]])
  }
  cat("Root attributes:", length(root_attrs), "keys copied\n")
}

# Known arrays in bcsd store
array_names <- c("latitude", "longitude", "time", "pr", "tas")

for (name in array_names) {
  v2_arr <- v2_root$get_item(name)
  data <- v2_arr$as.array()
  shape <- v2_arr$get_shape()
  chunks <- v2_arr$get_chunks()
  dtype <- v2_arr$get_dtype()
  fill_value <- v2_arr$get_fill_value()
  compressor <- v2_arr$get_compressor()

  cat(sprintf("  Array '%s': shape=(%s), dtype=%s, chunks=(%s)\n",
              name,
              paste(shape, collapse = ", "),
              dtype$dtype,
              paste(chunks, collapse = ", ")))

  # Create V3 array (zarr_format inherited from parent group)
  v3_arr <- v3_root$create_dataset(
    name = name,
    shape = as.integer(shape),
    chunks = as.integer(chunks),
    dtype = dtype$dtype,
    fill_value = fill_value,
    compressor = compressor
  )

  # Write data
  v3_arr$set_item("...", data)

  # Copy array attributes
  arr_attrs <- v2_arr$get_attrs()$to_list()
  if (length(arr_attrs) > 0) {
    v3_a <- v3_arr$get_attrs()
    for (nm in names(arr_attrs)) {
      v3_a$set_item(nm, arr_attrs[[nm]])
    }
    cat("    Attributes:", paste(names(arr_attrs), collapse = ", "), "\n")
  }
}

cat("\nV3 store written to:", v3_dir, "\n")

# Verify by reading back
cat("\nVerifying V3 store...\n")
v3_check_store <- DirectoryStore$new(v3_dir)
v3_check <- zarr_open_group(v3_check_store, mode = "r")

for (name in array_names) {
  arr <- v3_check$get_item(name)
  v2_arr <- v2_root$get_item(name)

  v3_data <- arr$as.array()
  v2_data <- v2_arr$as.array()

  shape_str <- paste(arr$get_shape(), collapse = ", ")
  match <- identical(dim(v3_data), dim(v2_data)) && all(v3_data == v2_data, na.rm = TRUE)
  cat(sprintf("  Verify '%s': shape=(%s), data_match=%s\n", name, shape_str, match))
}

# Zip it up
cat(sprintf("\nCreating %s...\n", v3_zip))
old_wd <- setwd(dirname(v3_dir))
zip(v3_zip, files = basename(v3_dir))
setwd(old_wd)

cat("Done!\n")
